src/dataflow/src/render/top_k.rs

```diff
@@ -31,14 +29,84 @@ where
             order_key,
             limit,
             offset,
+            monotonic,
         } = relation_expr
         {
             let arity = input.arity();
-            use differential_dataflow::operators::reduce::Reduce;
-
-            // self.ensure_rendered(input, scope, worker_index);
             let (ok_input, err_input) = self.collection(input).unwrap();
 
+            // For monotonic inputs, we are able to retract inputs not produced as outputs.
+            if *monotonic {
+                use differential_dataflow::operators::iterate::Variable;
+                use differential_dataflow::operators::Consolidate;
+                let delay = std::time::Duration::from_nanos(10_000_000_000);
+                let retractions = Variable::new(&mut ok_input.scope(), delay.as_millis() as u64);
+                let thinned = ok_input.concat(&retractions.negate());
+                let result = build_topk(thinned, group_key, order_key, *offset, *limit, arity);
+                retractions.set(&ok_input.concat(&result.negate()).consolidate());
```

Comment:
is the consolidate on this line mandatory for correctness or is it a performance optimization (i know in recursive scopes you have to consolidate the recursive variable to get it to converge and not sure if theres a similar thing here)

Reply:
I think it is probably not necessary, and added it while debugging to make sure I was seeing what was actually happening. I'll try out removing it. It would be necessary if there was a compaction-free path through the dataflow from the variables use to its setting.